lisa/search_space.py

```diff
@@ -0,0 +1,464 @@
+        else:
+            assert isinstance(capability, list), f"actual: {type(capability)}"
+            result = self.max if self.max_inclusive else self.max - 1
+            for cap_item in capability:
+                temp_result = self.meet_result(cap_item)
+                if temp_result.supported:
+                    temp_min = self.get_min_value(cap_item)
+                    result = min(temp_min, result)
+
+        return result
+
+
+CountSpace = Union[int, List[IntRange], IntRange, None]
+
+
+@dataclass_json(letter_case=LetterCase.CAMEL)
+@dataclass
+class ListSpace(RequirementMixin, List[T_SEARCH_SPACE]):
+    space: CountSpace = field(default_factory=IntRange)
+
+    def __init__(
+        self,
+        space: Optional[CountSpace] = None,
+        items: Optional[Iterable[T_SEARCH_SPACE]] = None,
+    ) -> None:
+        if items:
+            self.extend(items)
+        if space is not None:
+            self.space = space
+
+    def __post_init__(self, *args: Any, **kwargs: Any) -> None:
+        if len(self) > 1:
+            assert self.space == len(self), (
+                f"multiple items({len(self)}) need same number of count_space, "
+                f"but it's {self.space}"
+            )
+
+    def meet_result(self, capability: Any) -> RequirementResult:
+        result = RequirementResult()
+        if capability:
+            assert isinstance(capability, ListSpace), f"actual: {type(capability)}"
+            # meet number requirement
+            if isinstance(self.space, int) and isinstance(capability.space, int):
+                if self.space > capability.space:
+                    result.add_reason(
+                        f"capability's count({capability.space}) "
+                        f"must be more than requirement({self.space})"
+                    )
+            else:
+                result.merge(
+                    meet_result_countspace(self.space, capability.space),
+                    name="count_space",
+                )
+            if result.supported and not capability:
+                result.add_reason("capability should have at least one item")
+            if result.supported:
+                assert capability
+                if len(self) == 1:
+                    # if it's here, it means count already meet requirement, and
+                    # 1. both requirement and capability have exactly one.
+                    # 1. requirement need exact one, capability has more.
+                    # 2. requirement need at least one, capability has any number.
+                    # 3. requirement need at least one, capability has a range.
+                    item = self[0]
+                    for index in range(len(capability)):
+                        result.merge(
+                            item.meet_result(capability[index]), name=f"{index}"
+                        )
+                        if self.space == 1:
+                            break
```

Comment:
Using `enumerate()` here would be a good idea, too

Reply:
fixed